server/TypeDBServer.java

```diff
@@ -210,66 +226,67 @@ protected void serve() {
-            server.awaitTermination();
-            databaseMgr.close();
-            System.runFinalization();
-            logger().info("{} has been shutdown", name());
-        } catch (InterruptedException e) {
-            logger().error(FAILED_AT_STOPPING.message(), e);
-            Thread.currentThread().interrupt();
+    public synchronized void close() throws TypeDBCheckedException {
+        if (isOpen.compareAndSet(true, false)) {
+            try {
+                logger().info("");
+                logger().info("Closing {} instance...", name());
+                assert typeDBService != null;
+                typeDBService.close();
+                logger().info("Stopping network layer...");
+                server.shutdown();
+                if (!server.awaitTermination(10, TimeUnit.SECONDS)) {
+                    server.shutdownNow();
+                }
+                logger().info("Stopping storage layer...");
+                databaseMgr.close();
+                System.runFinalization();
+                logger().info("{} instance has been closed.", name());
+            } catch (Throwable e) {
+                logger().error(FAILED_AT_STOPPING.message(), e);
+                throw TypeDBCheckedException.of(e);
+            }
         }
     }
 
-    public static void main(String[] args) {
-        try {
-            printASCIILogo();
-
-            CoreConfigParser configParser = new CoreConfigParser();
-            ArgsParser<CoreSubcommand> argsParser = new ArgsParser<CoreSubcommand>()
-                    .subcommand(new CoreSubcommandParser.Server(configParser))
-                    .subcommand(new CoreSubcommandParser.Import())
-                    .subcommand(new CoreSubcommandParser.Export());
-            Optional<CoreSubcommand> subcmd = argsParser.parse(args);
-            if (subcmd.isEmpty()) {
-                LOG.error(UNRECOGNISED_CLI_COMMAND.message(String.join(" ", args)));
-                LOG.error(argsParser.usage());
-                System.exit(1);
-            } else {
-                if (subcmd.get().isServer()) {
-                    CoreSubcommand.Server subcmdServer = subcmd.get().asServer();
-                    if (subcmdServer.isHelp()) System.out.println(argsParser.help());
-                    else if (subcmdServer.isVersion()) System.out.println("Version: " + Version.VERSION);
-                    else runServer(subcmdServer);
-                } else if (subcmd.get().isImport()) {
-                    runImport(subcmd.get().asImport());
-                } else if (subcmd.get().isExport()) {
-                    runExport(subcmd.get().asExport());
-                } else throw TypeDBException.of(ILLEGAL_STATE);
-            }
-        } catch (Exception e) {
-            if (e instanceof TypeDBException) {
-                LOG.error(e.getMessage());
-            } else {
-                LOG.error(e.getMessage(), e);
-                LOG.error(EXITED_WITH_ERROR.message());
```

Comment:
we now expect every branch of this `if` block to handle their own exits and exit codes and never fail